### PR TITLE
FW-415. Fix.

### PR DESCRIPTION
**Pull request by changtengfei**
Main changes:

* implemented a sniffer base on radio project in OpenWSN
* make parameter of network configurable: like the period of EB, KA, DIO and DAO, whether ackReply, whether channel hopping enabled...
* add conditional compilation: `GOLDEN_IMAGE`
    * When `GOLDEN_IMAGE` defined, slot duration set to 10ms
    * When `GOLDEN_IMAGE` not defined, slot duration set to 15ms

### CODE REVIEW

**Comment by twatteyne** [comment]
wow, very cool!
Let me create a issue so we can track this correctly. In the future, let's try not to forget.

File: SConstruct [hunk 1]
```diff
@@ -133,7 +134,8 @@ command_line_options = {
     'debug':            ['0','1'],
     'noadaptivesync':   ['0','1'],
     'cryptoengine':     ['', 'dummy_crypto_engine', 'firmware_crypto_engine', 'board_crypto_engine'],
-    'l2_security':      ['0','1']
+    'l2_security':      ['0','1'],
+    'goldenImage':      ['none','root','sniffer']
```
Comment (twatteyne) [inline SConstruct]: add final `,`

File: SConstruct [hunk 2]
```diff
@@ -279,6 +281,13 @@ command_line_vars.AddVariables(
         int,                                               # converter
     ),
     (
+        'goldenImage',                                     # key
+        '',                                                # help
```
Comment (twatteyne) [inline SConstruct]: add comment "create an golden image for interop testing"

File: bsp/boards/OpenMote-CC2538/board_info.h [hunk 3]
```diff
@@ -66,9 +83,10 @@
 #define PORT_maxRxDataPrepare               33    // 1007us (measured  84us)
 #define PORT_maxTxAckPrepare                22    //  305us (measured 219us)
 // radio speed related
-#define PORT_delayTx                        7     //  214us (measured 219us)
+#define PORT_delayTx                        12    //  214us (measured 219us)
```
Comment (twatteyne) [inline bsp/boards/OpenMote-CC2538/board_info.h]: why is this different even if not `GOLDEN_IMAGE_ROOT`. Was this a bug?

File: drivers/common/openserial.c [hunk 4]
```diff
@@ -158,6 +162,27 @@ owerror_t openserial_printData(uint8_t* buffer, uint8_t length) {
...
 }
 
+owerror_t openserial_printPacket(uint8_t* buffer, uint8_t length, uint8_t channel) {
+   uint8_t  i;
+   INTERRUPT_DECLARATION();
+   
+   DISABLE_INTERRUPTS();
+   openserial_vars.outputBufFilled  = TRUE;
+   outputHdlcOpen();
+   outputHdlcWrite(SERFRAME_MOTE2PC_PACKET);
```
Comment (twatteyne) [inline drivers/common/openserial.c]: `SERFRAME_MOTE2PC_PACKET` -> `SERFRAME_MOTE2PC_SNIFFED_PACKET`

**Comment by twatteyne** [comment]
Please do not commit the `images/` directory (add to `.gitignore`)

File: openstack/02a-MAClow/IEEE802154E.c [hunk 5]
```diff
@@ -2037,9 +2082,14 @@ different channel offsets in the same slot.
 \returns The calculated frequency channel, an integer between 11 and 26.
 */
 port_INLINE uint8_t calculateFrequency(uint8_t channelOffset) {
-   // comment the following line out to disable channel hopping
-   return SYNCHRONIZING_CHANNEL; // single channel
-   //return 11+(ieee154e_vars.asnOffset+channelOffset)%16; //channel hopping
+    // comment the following line out to disable channel hopping
```
Comment (twatteyne) [inline openstack/02a-MAClow/IEEE802154E.c]: remove this line

File: openstack/02b-MAChigh/schedule.c [hunk 6]
```diff
@@ -35,6 +35,7 @@ void schedule_init() {
    }
    schedule_vars.backoffExponent = MINBE-1;
    schedule_vars.maxActiveSlots = MAXACTIVESLOTS;
+   schedule_vars.customFrameLength = SUPERFRAME_LENGTH;
```
Comment (twatteyne) [inline openstack/02b-MAChigh/schedule.c]: why "custom" frame length? Please replace by `frameLength`

Comment (twatteyne) [inline openstack/02b-MAChigh/schedule.c]: `SUPERFRAME_LENGTH` -> `SLOTRAME_LENGTH`

File: openstack/02b-MAChigh/schedule.c [hunk 7]
```diff
@@ -178,6 +179,15 @@ void schedule_setFrameLength(frameLength_t newFrameLength) {
    ENABLE_INTERRUPTS();
 }
 
+void schedule_setCustomFrameLength(frameLength_t newFrameLength) {
```
Comment (twatteyne) [inline openstack/02b-MAChigh/schedule.c]: `schedule_setCustomFrameLength` -> `schedule_setSlotframeLength`

File: openstack/02b-MAChigh/sixtop.c [hunk 8]
```diff
@@ -123,6 +123,7 @@ void sixtop_init() {
    sixtop_vars.dsn                = 0;
    sixtop_vars.mgtTaskCounter     = 0;
    sixtop_vars.kaPeriod           = MAXKAPERIOD;
+   sixtop_vars.ebPeriod           = EBTIMEOUT;
```
Comment (twatteyne) [inline openstack/02b-MAChigh/sixtop.c]: `EBTIMEOUT` -> `EBPERIOD`

File: openstack/03b-IPv6/icmpv6rpl.c [hunk 9]
```diff
@@ -269,7 +269,7 @@ void icmpv6rpl_timer_DIO_task() {
       sendDIO();
       
       // pick a new pseudo-random periodDIO
-      icmpv6rpl_vars.periodDIO = TIMER_DIO_TIMEOUT+(openrandom_get16b()&0xff);
```
Comment (twatteyne) [inline openstack/03b-IPv6/icmpv6rpl.c]: please remove rather than comment

File: openstack/03b-IPv6/icmpv6rpl.c [hunk 10]
```diff
@@ -394,7 +394,7 @@ void icmpv6rpl_timer_DAO_task() {
       sendDAO();
       
       // pick a new pseudo-random periodDAO
-      icmpv6rpl_vars.periodDAO = TIMER_DAO_TIMEOUT+(openrandom_get16b()&0xff);
```
Comment (twatteyne) [inline openstack/03b-IPv6/icmpv6rpl.c]: please remove rather than comment

File: openstack/03b-IPv6/icmpv6rpl.c [hunk 11]
```diff
@@ -567,3 +567,21 @@ void sendDAO() {
       openqueue_freePacketBuffer(msg);
    }
 }
+
+void icmpv6rpl_setDIOPeriod(uint16_t dioPeriod){
+    icmpv6rpl_vars.periodDIO = dioPeriod/5; // delayDIO will skip 4 times in 5 dio period
```
Comment (twatteyne) [inline openstack/03b-IPv6/icmpv6rpl.c]: I don't like this. 5 is a magic value. What if delayDIO does NOT skip. Please set `icmpv6rpl_vars.periodDIO` to `dioPeriod` and do something in `delayDIO` so it doesn't skip.

File: openstack/03b-IPv6/icmpv6rpl.c [hunk 12]
```diff
@@ -567,3 +567,21 @@ void sendDAO() {
...
+    opentimers_setPeriod(
+        icmpv6rpl_vars.timerIdDIO,
+        TIME_MS,
+        icmpv6rpl_vars.periodDIO
+     );
+}
+
+void icmpv6rpl_setDAOPeriod(uint16_t daoPeriod){
+    icmpv6rpl_vars.periodDAO = daoPeriod/5; // delayDAO will skip 4 times in 5 dao period
```
Comment (twatteyne) [inline openstack/03b-IPv6/icmpv6rpl.c]: same here, see comment for `icmpv6rpl_setDIOPeriod`

File: projects/common/03oos_sniffer/03oos_sniffer.c [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+   radio_rxEnable();
+   radio_rxNow();
+   
+   scheduler_start();
+   
+   return 0;
+}
+
+//=========================== interface =======================================
+void sniffer_setListeningChannel(uint8_t channel){
```
Comment (twatteyne) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: add a newline

File: projects/common/03oos_sniffer/03oos_sniffer.c [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+   // schedule again
+   bsp_timer_scheduleIn(TIMER_PERIOD);
+}
+
+// ================================ task =======================================
+void task_uploadPacket(){
+    openserial_printPacket(&(app_vars.packet[0]),app_vars.packet_len,app_vars.channel);
+}
+// ================================= stubbing ==================================
+void openbridge_triggerData(){return;}
```
Comment (twatteyne) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: add newline

File: projects/common/03oos_sniffer/03oos_sniffer.c [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+   } else {
+       openserial_stop();
+       openserial_startInput();
+   }
+   // schedule again
+   bsp_timer_scheduleIn(TIMER_PERIOD);
+}
+
+// ================================ task =======================================
+void task_uploadPacket(){
```
Comment (twatteyne) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: add newline

File: projects/common/03oos_sniffer/03oos_sniffer.c [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    openserial_printPacket(&(app_vars.packet[0]),app_vars.packet_len,app_vars.channel);
+}
+// ================================= stubbing ==================================
```
Comment (twatteyne) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: add newline

File: projects/common/03oos_sniffer/03oos_sniffer.h [hunk 17]
```diff
@@ -0,0 +1,15 @@
+/**
+\brief This program shows the use of the "sniffer" module.
+
+\author Tengfei Chang <tengfei.chang@eecs.berkeley.edu>, June 2015.
+*/
+
+#ifndef __openwsn_H
```
Comment (twatteyne) [inline projects/common/03oos_sniffer/03oos_sniffer.h]: wrong condition, should be `#ifndef __03OOS_SNIFFER_H`

**Comment by twatteyne** [comment]
very cool. Some very small suggestion before merging.

File: bsp/boards/OpenMote-CC2538/board_info.h (hunk 3 above)
Comment (changtengfei) [inline bsp/boards/OpenMote-CC2538/board_info.h]: Yes, there is something I need to say here. 12 is the right delayTx value for OpenMote-CC2538. With 12, OpenMote-CC2538 can synchronize perfectly.

When security enabled with 10 ms slot timing, the interval from received a frame to send back ack is not long enough for decryption of income packet and encrytion of ACK if the frame is long, like over 100 bytes. 

Setting delayTx smaller, will give some space for mote to processing long frame when security enabled with 10ms slot. However, the cost is we lost some accuracy on synchronization.

File: drivers/common/openserial.c (hunk 4 above)
Comment (changtengfei) [inline drivers/common/openserial.c]: Fixed.

File: openstack/02a-MAClow/IEEE802154E.c (hunk 5 above)
Comment (changtengfei) [inline openstack/02a-MAClow/IEEE802154E.c]: Fixed.

File: openstack/02b-MAChigh/schedule.c (hunk 7 above)
Comment (changtengfei) [inline openstack/02b-MAChigh/schedule.c]: `schedule_setFrameLength` is the default function provided by schedule component. I remove the schedule_setCustomFrameLength  function and using schedule_setFrameLength to do this job. See: https://github.com/changtengfei/openwsn-fw/commit/555cee397933954044a9dc524293f1d2fe45283e

File: openstack/02b-MAChigh/sixtop.c (hunk 8 above)
Comment (changtengfei) [inline openstack/02b-MAChigh/sixtop.c]: Fixed.

File: openstack/03b-IPv6/icmpv6rpl.c (hunk 9 above)
Comment (changtengfei) [inline openstack/03b-IPv6/icmpv6rpl.c]: Fixed.

File: openstack/03b-IPv6/icmpv6rpl.c (hunk 10 above)
Comment (changtengfei) [inline openstack/03b-IPv6/icmpv6rpl.c]: Fixed.

File: openstack/03b-IPv6/icmpv6rpl.c (hunk 11 above)
Comment (changtengfei) [inline openstack/03b-IPv6/icmpv6rpl.c]: Fixed by call DIO or DAO every `dioPeriod` and `daoPeriod`, without `delayDIO` or `delayDAO`

File: openstack/03b-IPv6/icmpv6rpl.c (hunk 12 above)
Comment (changtengfei) [inline openstack/03b-IPv6/icmpv6rpl.c]: Fixed.

File: projects/common/03oos_sniffer/03oos_sniffer.c (hunk 13 above)
Comment (changtengfei) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: Fixed.

File: projects/common/03oos_sniffer/03oos_sniffer.c (hunk 14 above)
Comment (changtengfei) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: Fixed.

File: projects/common/03oos_sniffer/03oos_sniffer.c (hunk 15 above)
Comment (changtengfei) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: Fixed.

File: projects/common/03oos_sniffer/03oos_sniffer.c (hunk 16 above)
Comment (changtengfei) [inline projects/common/03oos_sniffer/03oos_sniffer.c]: Fixed.

File: projects/common/03oos_sniffer/03oos_sniffer.h (hunk 17 above)
Comment (changtengfei) [inline projects/common/03oos_sniffer/03oos_sniffer.h]: Fixed.

File: bsp/boards/OpenMote-CC2538/board_info.h (hunk 3 above)
Comment (twatteyne) [inline bsp/boards/OpenMote-CC2538/board_info.h]: @changtengfei , I question whether setting `delayTx` is the right parameter. The idea of `delayTx` is to represent the delay between the moment the CPU say "GO, start transmitting" and the moment the SFD leaves the radio. The value of `delayTx` is hence only related to the speed of the hardware, and should have no relationship with e.g. the length of a slot.

To change the duration between a data packet and an ACK, why don't you modify `TsTxAckDelay`?

**Pull request merged**